src/util/mod.rs

```diff
@@ -99,3 +99,92 @@ pub struct BufferInitDescriptor<'a> {
     /// will panic.
     pub usage: crate::BufferUsage,
 }
+
+/// Utility methods not meant to be in the main API.
+pub trait QueueExt {
+    /// Upload an entire texture and its mipmaps from a source buffer.
+    ///
+    /// Expects all mipmaps to be tightly packed in the data buffer.
+    ///
+    /// If the texture is a 2DArray texture, uploads each layer in order, expecting
+    /// each layer and its mips to be tightly packed.
+    fn write_whole_texture_mips(
+        &self,
+        texture: MipmappedTextureCopyView,
+        data: &[u8],
+        size: crate::Extent3d,
+    );
+}
+
+impl QueueExt for crate::Queue {
+    fn write_whole_texture_mips(
+        &self,
+        texture: MipmappedTextureCopyView<'_>,
+        data: &[u8],
+        size: crate::Extent3d,
+    ) {
+        let format_info = texture.format.describe();
+
+        let (layer_iterations, mip_extent) =
+            if texture.texture_dimension == crate::TextureDimension::D3 {
+                (1, size)
+            } else {
+                (size.depth, crate::Extent3d { depth: 1, ..size })
+            };
+
+        let mut binary_offset = 0;
+        for layer in 0..layer_iterations {
```

Comment:
one important question here is - what guides the decision on the order between mipmap levels and array layers? I.e. why does this code iterate the layers in the outer loop, and levels in the inner loop, specifically?